src/Common/src/Microsoft/Win32/SafeHandles/SafeHeapHandleCache.cs

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+
+namespace Microsoft.Win32.SafeHandles
+{
+    /// <summary>
+    /// Allows limited reuse of heap buffers to improve memory pressure. This cache does not ensure
+    /// that multiple copies of handles are not released back into the cache.
+    /// </summary>
+    internal sealed class SafeHeapHandleCache : IDisposable
+    {
+        private readonly ulong _minSize;
+        private readonly ulong _maxSize;
+
+        private LowLevelConcurrentStack<SafeHeapHandle> handleCache = new LowLevelConcurrentStack<SafeHeapHandle>();
```

Comment:
> swapping with Parallel.For with varying max conc

This is almost certainly always going to be satisfied the pool, no matter what you specify as the max concurrency level, because as written it's going to get throttled by the thread pool.  My pooling concerns are around the cases where it's not satisfied from the pool.

> at least twice as fast

I tried it out locally, under the same constraints I previous described.  I see a single-threaded 50% improvement in throughput, but a 25% regression in GCs.  The 50% throughput improvement is great, and I'm saying that doing all of these changes in one batch like this makes it very difficult for anyone else to understand what individual pieces led to that improvement and whether other changes along the way helped or detracted from that.  Makes it very difficult for me to provide feedback that isn't then dismissed as "well this all-told is faster so it's good".  The 25% regression in GCs concerns me as it represents something that under load can significantly impact overall application performance.